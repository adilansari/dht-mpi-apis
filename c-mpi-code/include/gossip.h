/*
 * (C) 2001 Clemson University and The University of Chicago
 *
 * See COPYING in top-level directory.
 */

/** \defgroup gossip gossip logging interface
 *
 * This is a basic application logging facility.  It uses printf style
 * formatting and provides several mechanisms for output.
 *
 * @{
 */

/** \file
 *
 *  Declarations for the gossip logging interface.
 */

#ifndef __GOSSIP_H
#define __GOSSIP_H

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 500
#endif

#include <errno.h>
#include <stdio.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <sys/time.h>
#ifdef HAVE_EXECINFO_H
#include <execinfo.h>
#endif
#ifdef __GOSSIP_ENABLE_MPI__
#include <mpi.h>
#endif

#ifndef __KERNEL__
#include <stdint.h>
#include <stdarg.h>
#include "syslog.h"
#endif
#include "cmpi-config.h"

// Masks...
#define MASK_ALL     ((uint64_t) -1)
#define MASK_NONE    ((uint64_t) 0)
#define MASK_DEBUG   ((uint64_t) 1 << 1)
#define MASK_MPIRPC  ((uint64_t) 1 << 2)
#define MASK_DHT     ((uint64_t) 1 << 3)
#define MASK_ADTS    ((uint64_t) 1 << 4)
#define MASK_DRIVER  ((uint64_t) 1 << 5)
#define MASK_CMPI    ((uint64_t) 1 << 6)
#define MASK_ARCH    ((uint64_t) 1 << 7)
#define MASK_STORAGE ((uint64_t) 1 << 8)
#define MASK_KDA     ((uint64_t) 1 << 9)
#define MASK_EXAMPLE ((uint64_t) 1 << 10)
#define MASK_CLIENT  ((uint64_t) 1 << 12)

/********************************************************************
 * Visible interface
 */

#define GOSSIP_BUF_SIZE 1024

/* what type of timestamp to place in msgs */
enum gossip_logstamp
{
    GOSSIP_LOGSTAMP_NONE = 0,
    GOSSIP_LOGSTAMP_USEC = 1,
    GOSSIP_LOGSTAMP_DATETIME = 2,
    GOSSIP_LOGSTAMP_THREAD = 3,
    GOSSIP_LOGSTAMP_MPI = 4,
};
#define GOSSIP_LOGSTAMP_DEFAULT GOSSIP_LOGSTAMP_MPI

#define gossip_debug_init() \
do {\
    char* __mask_str; \
    uint64_t __mask; \
    __mask_str = getenv("CMPI_DEBUG_MASK"); \
    __mask = (__mask_str ? gossip_set_debug_mask_human(__mask_str) : 0); \
    gossip_set_debug_mask(1, __mask); \
}while(0)

uint64_t gossip_set_debug_mask_human(
    const char *event_logging);

/* Keep a simplified version for the kmod */
#ifdef __KERNEL__

#ifdef GOSSIP_DISABLE_DEBUG
#define gossip_debug(mask, format, f...) do {} while(0)
#else
extern int gossip_debug_mask;

/* try to avoid function call overhead by checking masks in macro */
#define gossip_debug(mask, format, f...)                  \
do {                                                      \
    if (gossip_debug_mask & mask)                         \
    {                                                     \
        printk(format, ##f);                              \
    }                                                     \
} while(0)
#endif /* GOSSIP_DISABLE_DEBUG */

/* do file and line number printouts w/ the GNU preprocessor */
#define gossip_ldebug(mask, format, f...)                  \
do {                                                       \
    gossip_debug(mask, "%s: " format, __func__ , ##f); \
} while(0)

#define gossip_err printk
#define gossip_lerr(format, f...)                  \
do {                                               \
    gossip_err("%s line %d: " format, __FILE__ , __LINE__ , ##f); \
} while(0)

#else /* __KERNEL__ */

/* stdio is needed by gossip_debug_fp declaration for FILE* */
#include <stdio.h>

int gossip_enable_syslog(
    int priority);
int gossip_enable_stdout(
    void);
int gossip_enable_stderr(
    void);
int gossip_enable_file(
    const char *filename,
    const char *mode);
int gossip_disable(
    void);
int gossip_set_debug_mask(
    int debug_on,
    uint64_t mask);
int gossip_get_debug_mask(
    int *debug_on,
    uint64_t *mask);
int gossip_set_logstamp(
    enum gossip_logstamp ts);

void gossip_backtrace(void);

#ifdef __GNUC__

/* do printf style type checking if built with gcc */
int __gossip_debug(
    uint64_t mask,
    char prefix,
    const char *format,
    ...) __attribute__ ((format(printf, 3, 4)));
int gossip_err(
    const char *format,
    ...) __attribute__ ((format(printf, 1, 2)));
int __gossip_debug_va(
    uint64_t mask,
    char prefix,
    const char *format,
    va_list ap);
int gossip_debug_fp(
    FILE *fp,
    char prefix,
    enum gossip_logstamp ts,
    const char *format,
    ...) __attribute__ ((format(printf, 4, 5)));

#ifdef GOSSIP_DISABLE_DEBUG
#define gossip_debug(mask, format, f...) do {} while(0)
#define gossip_perf_log(format, f...) do {} while(0)
#define gossip_debug_enabled(__m) 0
#else
extern int gossip_debug_on;
extern int gossip_facility;
extern uint64_t gossip_debug_mask;

#define gossip_debug_enabled(__m) \
    (gossip_debug_on && (gossip_debug_mask & __m))

/* try to avoid function call overhead by checking masks in macro */
#define gossip_debug(mask, format, f...)                  \
do {                                                      \
    if ((gossip_debug_on) && (gossip_debug_mask & mask) &&\
        (gossip_facility))                                \
    {                                                     \
        __gossip_debug(mask, '?', format, ##f);           \
    }                                                     \
} while(0)
#define gossip_perf_log(format, f...)                     \
do {                                                      \
    if ((gossip_debug_on) &&                              \
        (gossip_debug_mask & GOSSIP_PERFCOUNTER_DEBUG) && \
        (gossip_facility))                                \
    {                                                     \
        __gossip_debug(GOSSIP_PERFCOUNTER_DEBUG, 'P',     \
            format, ##f);                                 \
    }                                                     \
} while(0)
#define gossip_do(mask, x)                                      \
    if ((gossip_debug_on) && (gossip_debug_mask & mask) &&      \
        (gossip_facility))                                      \
    { x; }

#endif /* GOSSIP_DISABLE_DEBUG */

/* do file and line number printouts w/ the GNU preprocessor */
#define gossip_ldebug(mask, format, f...)                  \
do {                                                       \
    gossip_debug(mask, "%s %s() line %d: " format, __FILE__, __func__ , __LINE__, ##f); \
} while(0)

#define gossip_lerr(format, f...)                  \
do {                                               \
    gossip_err("%s %s() line %d: " format, __FILE__ , __func__, __LINE__ , ##f); \
    gossip_backtrace();                            \
} while(0)
#else /* ! __GNUC__ */

int __gossip_debug(
    uint64_t mask,
    char prefix,
    const char *format,
    ...);
int __gossip_debug_stub(
    uint64_t mask,
    char prefix,
    const char *format,
    ...);
int gossip_err(
    const char *format,
    ...);

#ifdef GOSSIP_DISABLE_DEBUG
#define gossip_debug(__m, __f, f...) __gossip_debug_stub(__m, '?', __f, ##f);
#define gossip_ldebug(__m, __f, f...) __gossip_debug_stub(__m, '?', __f, ##f);
#define gossip_debug_enabled(__m) 0
#else
#define gossip_debug(__m, __f, f...) __gossip_debug(__m, '?', __f, ##f);
#define gossip_ldebug(__m, __f, f...) __gossip_debug(__m, '?', __f, ##f);
#define gossip_debug_enabled(__m) \
            ((gossip_debug_on != 0) && (__m & gossip_debug_mask))

#endif /* GOSSIP_DISABLE_DEBUG */

#define gossip_lerr gossip_err

#endif /* __GNUC__ */

#endif /* __KERNEL__ */

#endif /* __GOSSIP_H */

/* @} */

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
