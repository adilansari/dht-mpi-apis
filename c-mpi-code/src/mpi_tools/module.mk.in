
DIR := src/mpi_tools

MPITOOLS_SRC :=
MPITOOLS_SRC += $(DIR)/mpi_tools.c
MPITOOLS_SRC += $(DIR)/io_tools.c

MPITOOLS_OBJS = $(patsubst %.c, %.o, $(MPITOOLS_SRC))
