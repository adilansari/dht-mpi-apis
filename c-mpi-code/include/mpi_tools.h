
#ifndef MPI_TOOLS
#define MPI_TOOLS

#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/time.h>

#include <mpi.h>

#ifdef ENABLE_DEBUG
#ifdef HAVE_EXEC_INFO_H
#include <execinfo.h>
#endif
#endif

#include <gossip.h>

extern int mpi_rank;
extern int mpi_size;

extern int debug_rank;

extern FILE* debug_file;

/**
   Quick setup
 */
void whoami(void);

char* heap(char* s);
char* iheap(int i);
char* xheap(int i);

/**
   Execute code line or block in rank order.
*/
#define sequential(code)                                                \
  {                                                                     \
    int sequential_i;                                                   \
    for (sequential_i = 0; sequential_i < mpi_size;                     \
         sequential_i++)                                                \
    {                                                                   \
      if (mpi_rank == sequential_i)                                     \
      {                                                                 \
        code;                                                           \
      }                                                                 \
      MPI_Barrier(MPI_COMM_WORLD);                                      \
    }                                                                   \
  }

void printbinary(unsigned int number);
void printdata(char* tag, char* data, int length);
void fprintdata(FILE* file, char* tag, char* data, int length);
int  flip2(int i);
unsigned int ilog2(unsigned int j);
unsigned int isqrt(unsigned int j);

/**
   Return a random number in [0,n).
*/
int rand_lt(int n);

/**
   Return a random double in [0,n).
*/
double random_double(int n);

/**
   Concatenate char* s at char* p, update p to the end.
*/
#define cat(p, s) p += sprintf(p, "%s", s);

#ifdef ENABLE_DEBUG
// #warning "ENABLE_DEBUG"
#define DEBUG(x) x
#else
#define DEBUG(x)
#endif

#ifdef ENABLE_DEBUG
/**
   Dump the stack and exit.
*/
void
dump_stack(void);
#endif

#define OK NOTE("ok")

void setup_debug_file(void);

/**
   Simplify debugging statements.
 */
void note(char* msg);
void note_b(char* msg, int b);
void note_d(char* msg, double d);
void note_i(char* msg, int i);
void note_ii(char* msg, int i, int j);
void note_s(char* msg, char* s);
void note_ss(char* msg, char* s1, char* s2);
void note_p(char* msg, void* p);
void note_u(char* msg, unsigned int u);
void note_x(char* msg, int i);
void note_xi(char* msg, int i, int j);
void note_sx(char* msg, char* s, int i);
void note_sxi(char* msg, char* s, int x, int i);
// void note_ssxi(char* msg, char* s, int x, int i);

void show_b(char* tag, int b);
void show_d(char* tag, double d);
void show_i(char* tag, int i);
void show_ii(char* tag_i, int i, char* tag_j, int j);
void show_l(char* tag, long l);
void show_s(char* tag, char* s);
void show_u(char* tag, unsigned int u);
void show_x(char* tag, int x);
void show_p(char* tag, void* p);

void  timestamp(char* label, char* msg);
char* timestring(char* label, char* msg);

#define NOTE(m)                DEBUG(note(m))
#define NOTE_B(m,b)            DEBUG(note_b(m,b))
#define NOTE_D(m,d)            DEBUG(note_d(m,d))
#define NOTE_I(m,i)            DEBUG(note_i(m,i))
#define NOTE_II(m,i,j)         DEBUG(note_ii(m,i,j))
#define NOTE_U(m,u)            DEBUG(note_u(m,u))
#define NOTE_P(m,s)            DEBUG(note_p(m,))
#define NOTE_S(m,s)            DEBUG(note_s(m,s))
#define NOTE_SX(m,s,i)         DEBUG(note_sx(m,s,i))
#define NOTE_SXI(m,s,x,i)      DEBUG(note_sxi(m,s,x,i))
// #define NOTE_SSXI(m,s1,s2,x,i) DEBUG(note_ssxi(m,s1,s2,x,i))
#define NOTE_X(m,i)            DEBUG(note_x(m,i))
#define NOTE_XI(m,i,j)         DEBUG(note_xi(m,i,j))

#define NOTE_F DEBUG(fprintf(debug_file, "[%i] %s()... \n",     \
                             debug_rank, __func__);             \
                     fflush(debug_file);)
#define DONE DEBUG(fprintf(debug_file, "[%i] %s() done. \n",     \
                          debug_rank, __func__);        \
                   fflush(debug_file);)
#define NOTE_FI(i)                              \
  DEBUG(fprintf(debug_file, "[%i] %s(%i)... \n",             \
               debug_rank, __func__, i);        \
        fflush(stdout);)
#define NOTE_FX(i) DEBUG(fprintf(debug_file, "[%i] %s(%X)... \n",            \
                                debug_rank, __func__, i);       \
                         fflush(stdout);)
#define NOTE_FIS(i,s) DEBUG(fprintf(debug_file, "[%i] %s(%i,%s)... \n",      \
                                   debug_rank, __func__, i, s); \
                            fflush(stdout);)
#define NOTE_FXIS(x,i,s)                        \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%i,%s)... \n",       \
               debug_rank, __func__, x, i, s);  \
        fflush(stdout);)
#define NOTE_FXI(i,j)                           \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%i)... \n",          \
               debug_rank, __func__, i, j);     \
        fflush(stdout);)
#define NOTE_FXS(i,s)                           \
  DEBUG(fprintf(debug_file, "[%i] %s(%X,%s)... \n",          \
               debug_rank, __func__, i, s);     \
        fflush(stdout););
#define NOTE_FXX(i,j) DEBUG(fprintf(debug_file, "[%i] %s(%X,%X)... \n",      \
                                   debug_rank, __func__, i, j); \
                            fflush(stdout);)
#define NOTE_FS(s) DEBUG(fprintf(debug_file, "[%i] %s(%s)... \n",            \
                                debug_rank, __func__, s);       \
                         fflush(stdout);)
#define NOTE_FSS(s1,s2)                         \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s)... \n",          \
               debug_rank, __func__, s1, s2);   \
        fflush(stdout);)
#define NOTE_FSSS(s1,s2,s3)                     \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s,%s)... \n",       \
               debug_rank, __func__,            \
               s1, s2, s3);                     \
        fflush(stdout);)
#define NOTE_FSSSI(s1,s2,s3,i)                          \
  DEBUG(fprintf(debug_file, "[%i] %s(%s,%s,%s,%i)... \n",            \
               debug_rank, __func__, s1, s2, s3, i);    \
        fflush(stdout);)

///// Simple variable debugging...
#define SHOW_B(b)    DEBUG(show_b(#b, b))
#define SHOW_D(d)    DEBUG(show_d(#d, d))
#define show_i(i)    note_i(#i ": ", i)
#define SHOW_I(i)    DEBUG(show_i(i))
#define SHOW_II(i,j) DEBUG(show_ii(#i, i, #j, j))
#define SHOW_L(l)    DEBUG(show_l(#l, l))
#define SHOW_U(u)    DEBUG(show_u(#u, u))
#define show_s(s)    note_s(#s ": ", s)
#define SHOW_S(s)    DEBUG(show_s(s))
#define SHOW_X(x)    DEBUG(show_x(#x, x))
#define SHOW_P(p)    DEBUG(show_p(#p, p))
#define SHOW         SHOW_S

//// Functions...
#define show_fi(i)                              \
       fprintf(debug_file, "[%i] %s(%s=%i)... \n",          \
               debug_rank, __func__, #i, i);    \
        fflush(stdout);
#define SHOW_FI(i) DEBUG(show_fi(i))
#define SHOW_FIS(i,s)                                   \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s)... \n",            \
               debug_rank, __func__, #i, i, #s, s);     \
        fflush(stdout);)
#define SHOW_FISS(i,s1,s2)                                      \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s,%s=%s)... \n",              \
               debug_rank, __func__, #i, i, #s1, s1, #s2, s2);  \
        fflush(stdout);)
#define SHOW_FISX(i,s,x)                                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s,%s=%X)... \n",         \
                debug_rank, __func__, #i, i, #s, s, #x, x);             \
        fflush(stdout);)

#define show_fs(s) fprintf(debug_file,           \
                           "[%i] %s(%s=%s)... \n",          \
               debug_rank, __func__, #s, s);    \
        fflush(stdout);
#define SHOW_FS(s)                              \
  DEBUG(show_fs(s))

#define SHOW_FX(x)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X)... \n",            \
               debug_rank, __func__, #x, x);      \
        fflush(stdout);)
#define SHOW_FII(i1,i2)                                     \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%i)... \n",                \
               debug_rank, __func__, #i1, i1, #i2, i2);     \
        fflush(stdout);)

#define show_fiii(i1,i2,i3)                              \
      fprintf(debug_file, "[%i] %s(%s=%i,%s=%i,%s=%i)... \n",  \
              debug_rank, __func__, #i1, i1, #i2, i2, #i3, i3);   \
        fflush(stdout);
#define SHOW_FIII(i1,i2,i3) DEBUG(show_fiii(i1,i2,i3))

#define show_fsi(s,i)  \
      fprintf(debug_file, "[%i] %s(%s=%s,%s=%i)... \n",  \
                debug_rank, __func__, #s, s, #i, i);     \
        fflush(stdout);
#define SHOW_FSI(s,i) DEBUG(show_fsi(s,i))

#define show_fsii(s,i1,i2)                                     \
  fprintf(debug_file, "[%i] %s(%s=%s,%s=%i,%s=%i)... \n",      \
              debug_rank, __func__, #s, s, #i1, i1, #i2, i2);  \
  fflush(stdout);
#define SHOW_FSII(s,i1,i2) DEBUG(show_fsii(s,i1,i2))

#define SHOW_FXI(x,i)                                   \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i)... \n",            \
               debug_rank, __func__, #x, x, #i, i);     \
        fflush(stdout);)

#define SHOW_FXIS(x,i,s)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i,%s=%s)... \n",      \
               debug_rank, __func__,                    \
               #x, x, #i, i, #s, s);                    \
        fflush(stdout);)

#define SHOW_FXIIS(x,i1,i2,s)                                \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i,%s=%i,%s=%s)... \n",      \
               debug_rank, __func__,                    \
               #x, x, #i1, i1, #i2, i2, #s, s);                    \
        fflush(stdout);)

#define SHOW_FSS(s1,s2)                                 \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s,%s=%s)... \n",            \
               debug_rank, __func__, #s1, s1, #s2, s2); \
        fflush(stdout);)
#define SHOW_FSSI(s1,s2,i)                                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s,%s=%s,%s=%i)... \n",        \
                debug_rank, __func__, #s1, s1, #s2, s2, #i, i);        \
        fflush(stdout);)
#define SHOW_FISSI(i1,s1,s2,i2)                                \
  DEBUG(fprintf(debug_file,                                        \
                "[%i] %s(%s=%i,%s=%s,%s=%s,%s=%i)... \n",    \
                debug_rank, __func__, #i1, i1, #s1, s1,   \
                #s2, s2, #i2, i2); \
        fflush(stdout);)
#define SHOW_FSISSI(s1,i1,s2,s3,i2)                                \
  DEBUG(fprintf(debug_file,                                        \
                "[%i] %s(%s=%s,%s=%i,%s=%s,%s=%s,%s=%i)... \n",    \
                debug_rank, __func__, #s1, s1, #i1, i1, #s2, s2,   \
                #s3, s3, #i2, i2); \
        fflush(stdout);)
#define SHOW_FP(p)                              \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%p)... \n",          \
               debug_rank, __func__, #p, p);    \
        fflush(stdout);)

//// Functions with return values...

#define SHOW_FS_X(s,x)                                                  \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%s) -> %X... \n",               \
                debug_rank, __func__, #s, s, x);                        \
        fflush(stdout);)
#define SHOW_FIFS_S(i,f,s1,s2)                                       \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s(%s) -> %s)... \n",     \
                debug_rank, __func__, #i, i, f, s1, s2);             \
        fflush(stdout);)
#define SHOW_FIS_X(i,s,x)                                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s=%s) -> %X... \n",       \
                debug_rank, __func__, #i, i, #s, s, x);               \
        fflush(stdout);)
#define SHOW_FXI_B(x,i,b)                                               \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%X,%s=%i) -> %s... \n",         \
                debug_rank, __func__,                                   \
                #x, x, #i, i, b ? "true" : "false");                    \
        fflush(stdout);)


//// Functions of functions...

#define SHOW_FIF(i, f)                               \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s())... \n",           \
               debug_rank, __func__, #i, i, f);       \
        fflush(stdout);)
#define SHOW_FIFS(i, f, s)                             \
  DEBUG(fprintf(debug_file, "[%i] %s(%s=%i,%s(%s))... \n",           \
               debug_rank, __func__, #i, i, f, s);      \
        fflush(stdout);)

#define TIMESTAMP(m) note_i(m, time());

#define DUMPSTACK DEBUG(dump_stack());

char* file_string(char* file);

//// Memory usage debugging:
#ifdef DMALLOC
#include <dmalloc.h>
#define DMALLOC_SETUP(x) x
void dmalloc_setup(void);
#else
#define DMALLOC_SETUP(x)
#define dmalloc_setup(x)
#endif

#define DMALLOC_COMPLETE                              \
  DMALLOC_SETUP(                                      \
    {                                                 \
      SHOW_S(dmalloc_logpath);                        \
      dmalloc_shutdown();                             \
      dmalloc_log_unfreed();                          \
    });

/**
   Return a number in [0,n-1].
*/
int roll(int n);

#endif
