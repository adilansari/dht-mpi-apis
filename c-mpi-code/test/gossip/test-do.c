
/**
 * Test gossip_do() functionality.
 * */

#include <mpi.h>
#include <gossip.h>
#include <mpi_tools.h>

int main(int argc, char* argv[])
{
    MPI_Init(&argc, &argv);

    whoami();

    printf(
"This test program will print an example debugging message on the final\n"
"line of output if the CMPI_DEBUG_MASK environment variable is set to include\n"
"\"example\" or \"all\" in its comma separated list of masks:\n"
    );

    gossip_debug_init();

    gossip_do(MASK_EXAMPLE, printf("EXAMPLE MESSAGE\n"));

    MPI_Finalize();
    return 0;
}

/*
 * Local variables:
 *  c-indent-level: 2
 *  c-basic-offset: 2
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
