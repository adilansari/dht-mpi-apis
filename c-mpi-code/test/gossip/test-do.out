Debug to stdout
This test program will print an example debugging message on the final
line of output if the CMPI_DEBUG_MASK environment variable is set to include
"example" or "all" in its comma separated list of masks:
Debug to stdout
This test program will print an example debugging message on the final
line of output if the CMPI_DEBUG_MASK environment variable is set to include
"example" or "all" in its comma separated list of masks:
EXAMPLE MESSAGE
Debug to stdout
This test program will print an example debugging message on the final
line of output if the CMPI_DEBUG_MASK environment variable is set to include
"example" or "all" in its comma separated list of masks:
EXAMPLE MESSAGE
